File: src/Parameters/Input/InputParams.cpp
#include "InputParams.h"

#include "../Vars/Vector.h"
#include "../../Func/Func.h"

InputParams::InputParams() {}

InputParams::~InputParams() {
    delete x1;
    delete p;
    delete alfa;
    delete y;
}

std::string *InputParams::toString() {
    char number[15];
    std::string *str = new std::string("Input Parameters:\n");

    sprintf(number, "\tm: %d;\n", m);
    (*str) += number;

    sprintf(number, "\tx1: %s;\n", x1->toString()->c_str());
    (*str) += number;

    (*str) += "\tp:\n\t\t";
    (*str) += *(p->toString());

    (*str) += "\talfa:\n\t\t";
    (*str) += *(alfa->toString());

    sprintf(number, "\talfa_h: %0.10f;\n", alfa_h);
    (*str) += number;

    sprintf(number, "\teps_arg: %0.10f;\n", eps_arg);
    (*str) += number;

    sprintf(number, "\teps_f: %0.10f;\n", eps_f);
    (*str) += number;

    (*str) += "Function: " + y->func + ";\n";

    return str;
}

InputParams *InputParams::clone() {
    InputParams *input = new InputParams();
    input->m = m;
    input->x1 = x1 != nullptr ? x1->clone() : x1;
    input->p = p != nullptr ? p->clone() : p;
    input->alfa = alfa != nullptr ? alfa->clone() : alfa;
    input->alfa_h = alfa_h;
    input->eps_arg = eps_arg;
    input->eps_f = eps_f;
    input->y = y != nullptr ? y->clone() : y;

    return input;
}

File: Parametrs/InputParams.cs
using System;
using MethodsOptimization.src.Parametrs.Vars;
using MethodsOptimization.src.Functions;

namespace MethodsOptimization.src.Parametrs
{
    /// <summary>
    /// Входные данные
    /// </summary>
    class InputParams: ICloneable
    {
        public InputParams() {}
        
        /// <summary>
        /// стартовая точка
        /// </summary>
        public Vector X0 { get; set; }
        /// <summary>
        /// начальное направление
        /// </summary>
        public Vector P { get; set; }
        /// <summary>
        /// начальный интервал локализации минимума
        /// </summary>
        public Vector Alfa { get; set; }
        /// <summary>
        /// шаг изменения интервала локализации
        /// </summary>
        public double Alfa_h { get; set; } = 1e-2;
        /// <summary>
        /// Коэффициент ускорения
        /// </summary>
        public double B { get; set; } = 10;
        /// <summary>
        /// Иследуемая функция
        /// </summary>
        public Function Y { get; set; }
        /// <summary>
        /// Ограничения
        /// </summary>
        public LimitingParams Lim { get; set; }

        /// <summary>
        /// Привести к строке
        /// </summary>
        /// <returns></returns>
        new public string ToString()
        {
            string str = "Input:\n";
            if (X0 != null) str += "\tX0: " + X0.ToString().Replace("\t", "\t\t");
            if (P != null) str += "\tP: " + P.ToString().Replace("\t", "\t\t");
            if (Alfa != null) str += "\tAlfa: " + Alfa.ToString().Replace("\t", "\t\t");
            str += "\tAlfa_h: " + Alfa_h + ";\n";
            str += "\tB: " + B + ";\n";
            if (Y != null) str += "\tFunction: " + Y.ToString().Replace("\t", "\t\t");
            if (Lim != null) str += "\tLimiting params: " + Lim.ToString().Replace("\t", "\t\t");
            return str;
        }

        /// <summary>
        /// Получить дубликат
        /// </summary>
        /// <returns></returns>
        public object Clone()
        {
            InputParams clone = new InputParams();
            clone.X0 = (Vector)X0.Clone();
            clone.P = (Vector)P.Clone();
            clone.Alfa = (Vector)Alfa.Clone();
            clone.Alfa_h = Alfa_h;
            clone.B = B;
            clone.Y = Y;
            clone.Lim = (LimitingParams)Lim.Clone();
            return clone;
        }
    }
}
